### PR TITLE
cephadm: Various properties like 'last_refresh' do not contain timezone 

### PR DESCRIPTION
**Pull request by votdev**
The property `last_refresh`, `created`, `started`, `last_configured` or `last_deployed` in `ServiceDescription` and `DaemonDescription` do not contain any timezone information. This is problematic because the Dashboard can not display the correct time when no time zone information is set. It would help to simply add `Z` to the ISO 8601 strings if it is in UTC. All dates in Ceph should use timezone information according to https://github.com/ceph/ceph/pull/27799.

Fixes: https://tracker.ceph.com/issues/48068

Signed-off-by: Volker Theile <vtheile@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by mgfritch — APPROVED** [review]

**Comment by bk201** [comment]
- CI branch: [wip-kiefer-testing-2020-11-04-1131](https://github.com/ceph/ceph-ci/commits/wip-kiefer-testing-2020-11-04-1131)
- [CI packages](https://shaman.ceph.com/builds/ceph/wip-kiefer-testing-2020-11-04-1131/)
- QA runs: **FAIL**
  Suite|Total|Pass|Fail|Dead
  -|-|-|-|-
  [rados:cephadm](https://pulpito.ceph.com/kchang-2020-11-04_06:38:14-rados:cephadm-wip-kiefer-testing-2020-11-04-1131-distro-basic-smithi)|56|52|3|1

**Comment by bk201** [comment]
We may need to consider the upgrade case. If I apply the code to a running cluster and reload the module, I'll get this health error:

```
  cluster:
    id:     4c7fa6f2-6815-45a9-9c8b-5826ace2efcf
    health: HEALTH_ERR
            Module 'cephadm' has failed: can't compare offset-naive and offset-aware datetimes
```

The traceback in the mgr log:
```
2020-11-05T07:27:58.566+0000 7faa422cc640  0 [cephadm ERROR cephadm.utils] executing refresh((['mgr0', 'osd0'],)) failed.
Traceback (most recent call last):
  File "/ceph/src/pybind/mgr/cephadm/utils.py", line 58, in do_work
    return f(*arg)
  File "/ceph/src/pybind/mgr/cephadm/serve.py", line 114, in refresh
    if self.mgr.cache.host_needs_check(host):
  File "/ceph/src/pybind/mgr/cephadm/inventory.py", line 471, in host_needs_check
    return host not in self.last_host_check or self.last_host_check[host] < cutoff
TypeError: can't compare offset-naive and offset-aware datetimes
```

Cephadm stores some offset-naive datetime in the config store, which casues the error above after applying the changes in this PR. Here is the dump:

```
# ceph config-key dump mgr/cephadm | grep 2020
...
\"last_daemon_update\": \"2020-11-05T07:27:07.117910\", \"last_device_update\": \"2020-11-05T07:27:33.269936\",
...
\"last_host_check\": \"2020-11-05T07:27:06.910983\"
```
Not sure if I miss anything.

**Comment by votdev** [comment]
> We may need to consider the upgrade case. If I apply the code to a running cluster and reload the module, I'll get this health error:
> 
> ```
>   cluster:
>     id:     4c7fa6f2-6815-45a9-9c8b-5826ace2efcf
>     health: HEALTH_ERR
>             Module 'cephadm' has failed: can't compare offset-naive and offset-aware datetimes
> ```
> 
> The traceback in the mgr log:
> 
> ```
> 2020-11-05T07:27:58.566+0000 7faa422cc640  0 [cephadm ERROR cephadm.utils] executing refresh((['mgr0', 'osd0'],)) failed.
> Traceback (most recent call last):
>   File "/ceph/src/pybind/mgr/cephadm/utils.py", line 58, in do_work
>     return f(*arg)
>   File "/ceph/src/pybind/mgr/cephadm/serve.py", line 114, in refresh
>     if self.mgr.cache.host_needs_check(host):
>   File "/ceph/src/pybind/mgr/cephadm/inventory.py", line 471, in host_needs_check
>     return host not in self.last_host_check or self.last_host_check[host] < cutoff
> TypeError: can't compare offset-naive and offset-aware datetimes
> ```
> 
> Cephadm stores some offset-naive datetime in the config store, which casues the error above after applying the changes in this PR. Here is the dump:
> 
> ```
> # ceph config-key dump mgr/cephadm | grep 2020
> ...
> \"last_daemon_update\": \"2020-11-05T07:27:07.117910\", \"last_device_update\": \"2020-11-05T07:27:33.269936\",
> ...
> \"last_host_check\": \"2020-11-05T07:27:06.910983\"
> ```
> 
> Not sure if I miss anything.

That's for the feedback. I realized that i missed to convert naive datetime objects. With the latest changes this has been fixed.

**Comment by votdev** [comment]
jenkins test make check

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -93,7 +92,7 @@ if sys.version_info > (3, 0):
 container_path = ''
 cached_stdin = None
 
-DATEFMT = '%Y-%m-%dT%H:%M:%S.%f'
+DATEFMT = '%Y-%m-%dT%H:%M:%S.%fZ'
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:95]: Note that we need to continue to parse dates without `Z`

File: src/pybind/mgr/cephadm/module.py [hunk 2]
```diff
@@ -87,8 +86,6 @@ def remoto_has_connection(self):
   ConnectTimeout=30
 """
 
-CEPH_DATEFMT = '%Y-%m-%dT%H:%M:%S.%fZ'
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:90]: Note that `CEPH_DATEFMT` is the format used in the monmap, not the format which is used for json generation.

File: src/pybind/mgr/cephadm/module.py [hunk 3]
```diff
@@ -472,9 +469,8 @@ def notify(self, notify_type, notify_id):
         if notify_type == "mon_map":
             # get monmap mtime so we can refresh configs when mons change
             monmap = self.get('mon_map')
-            self.last_monmap = datetime.datetime.strptime(
-                monmap['modified'], CEPH_DATEFMT)
-            if self.last_monmap and self.last_monmap > datetime.datetime.utcnow():
+            self.last_monmap = str_to_datetime(monmap['modified'])
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:472]: I mean, if it works, that's great!

File: src/pybind/mgr/cephadm/tests/fixtures.py [hunk 4]
```diff
@@ -55,7 +54,7 @@ def with_cephadm_module(module_options=None, store=None):
             store = {}
         if '_ceph_get/mon_map' not in store:
             m.mock_store_set('_ceph_get', 'mon_map', {
-                'modified': datetime.datetime.utcnow().strftime(CEPH_DATEFMT),
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/tests/fixtures.py:58]: we're mocking the mon_map here. I mean, if it works, that would be great!

**Review by sebastian-philipp — APPROVED** [review]

File: src/cephadm/cephadm (hunk 1 above)
Comment (votdev) [reply src/cephadm/cephadm:95]: Sure, the helper function [str_to_datetime](https://github.com/ceph/ceph/pull/37920/files#diff-601cdd0e9db99fa5fc831bebf2d13d4d85be9d68664ef091ddfee90fcf336beaR26) is able to do that.

File: src/pybind/mgr/cephadm/module.py (hunk 3 above)
Comment (votdev) [reply src/pybind/mgr/cephadm/module.py:472]: The function `str_to_datetime` can parse multiple formats that are used in Ceph now and in the past.

File: src/pybind/mgr/cephadm/tests/fixtures.py (hunk 4 above)
Comment (votdev) [reply src/pybind/mgr/cephadm/tests/fixtures.py:58]: The function `datetime_to_str` is using the format `%Y-%m-%dT%H:%M:%S.%fZ`, so everything should be fine.

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2020-12-11_18:42:23-rados:cephadm-wip-swagner-testing-2020-12-11-1812-distro-basic-smithi/

failed with

```
2020-12-11T19:03:11.472 INFO:teuthology.orchestra.run.smithi176.stderr:--> Zapping: /dev/vg_nvme/lv_4
2020-12-11T19:03:11.472 INFO:teuthology.orchestra.run.smithi176.stderr:Running command: /usr/bin/dd if=/dev/zero of=/dev/vg_nvme/lv_4 bs=1M count=10 conv=fsync
2020-12-11T19:03:11.473 INFO:teuthology.orchestra.run.smithi176.stderr: stderr: 10+0 records in
2020-12-11T19:03:11.473 INFO:teuthology.orchestra.run.smithi176.stderr:10+0 records out
2020-12-11T19:03:11.473 INFO:teuthology.orchestra.run.smithi176.stderr: stderr: 10485760 bytes (10 MB, 10 MiB) copied, 0.0109941 s, 954 MB/s
2020-12-11T19:03:11.473 INFO:teuthology.orchestra.run.smithi176.stderr:--> Zapping successful for: <LV: /dev/vg_nvme/lv_4>
2020-12-11T19:03:12.015 DEBUG:teuthology.orchestra.run.smithi176:> sudo /home/ubuntu/cephtest/cephadm --image quay.ceph.io/ceph-ci/ceph:f8dbeb96eb0e35d1a67a073c06084411ff128972 shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid 21df8f04-3be3-11eb-9811-001a4aab830c -- ceph orch daemon add osd smithi176:vg_nvme/lv_4
```
this command ^^^ simply hangs now for hours.

No idea if this is caused by this one or other PRs.

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2021-01-07_14:03:38-rados:cephadm-wip-swagner-testing-2021-01-07-1213-distro-basic-smithi/

**Pull request merged**